src/controllers/productsController.ts

```diff
@@ -0,0 +1,99 @@
+import { Request, Response } from 'express';
+import logger from '../logs/config';
+import uploadImage from '../helpers/claudinary';
+import { validateFields } from '../validations';
+import { Size, SizeCreationAttributes } from '../database/models/Size';
+import { Product, ProductAttributes } from '../database/models/Product';
+
+export const createProduct = async (req: Request, res: Response) => {
+  try {
+    // when products exists
+    const name = req.body.name;
+    const thisProductExists = await Product.findOne({ where: { name } });
+
+    if (thisProductExists) {
+      return res.status(400).json({
+        ok: false,
+        message: 'This Product already exists, You can update the stock levels instead.',
+        data: thisProductExists,
+      });
+    }
+    // handle images
+    const productImages = [];
+    const images: unknown = req.files;
+    if (images instanceof Array && images.length > 3) {
+      for (const image of images) {
+        const imageBuffer: Buffer = image.buffer;
+        const url = await uploadImage(imageBuffer);
+        productImages.push(url);
+      }
+    } else {
+      return res.status(400).json({
+        message: 'Product should have at least 4 images',
+      });
+    }
+
+    // Create product
+    const categoryId = req.params.categoryId;
+    const requiredFields = ['name', 'description', 'sellerId'];
+    const missingFields = validateFields(req, requiredFields);
+
+    if (missingFields.length > 0) {
+      res.status(400).json({
+        ok: false,
+        message: `Required fields are missing: ${missingFields.join(', ')}`,
+      });
+    }
+
+    const { description, discount, sellerId, expiryDate, price, quantity } = req.body as ProductAttributes;
+
+    const createdProduct = await Product.create({
+      sellerId,
+      name,
+      description,
+      discount,
+      categoryId,
+      expiryDate,
+      price,
+      quantity,
+      images: productImages,
+    });
+
+    res.status(201).json({
+      ok: true,
+      data: createdProduct,
+      message: 'Thank you for adding new product in the store!',
+    });
+  } catch (error) {
+    if (error instanceof Error) {
```

Comment:
i'll consider this but i guess this makes our code to run slowly, cuz we already have express built-in way of handling errors, idk why we use another async function to do that

Reply:
No ,it doesn't the reason behind all that @Favor-star is telling is to use re-usable functions and reduce code duplication ,so it would be the best approach for sake of our code maintainability